core/domain/learner_progress_services.py

```diff
@@ -1042,7 +1042,9 @@ def _get_filtered_completed_story_summaries(
         else:
             story_id = story_summary.id
             story = stories[index]
-            # Ruling out the possibility of None for mypy type checking.
+            # Ruling out the possibility of None for mypy type checking, because
```

Comment:
Why don't you filter the list for None before you pass it in, so that the type can be made non-optional and you don't need the assert? That way, the list generation can be done closer to the location which provides info on whether the story exists or not.

Try to consider this sort of thing whenever you find you need to write an assertion and a long piece of reasoning justifying it.

Reply:
If I'm not wrong you mean we should do something like. below:
```
stories_with_none = story_fetchers.get_stories_by_ids(story_ids)
stories = [story for story in stories if story is not None ]
```
and then use it further in the code, right ? If not, please explain!

Reply:
No, that is not what I mean. I mean, why can't the callers of this function ensure that they pass in only items that are non-None, so that this function is guaranteed not to need to deal with this case?

Reply:
The main motive of this function is to filter the stories into different categories. So, in the argument `story_summaries` we are accepting `List[Optional[story_domain.StorySummary]` and these are the story_summaries of story_ids that were passed to the `story_ids` argument.
So, if a story_summary is None then we are putting that story's id into the `nonexistent_completed_story_ids` list (line 1041) and if a story_summary is None then it's a sure thing that it's corresponding story also does not exists. 
So, that is why assert is valid here. Because in this else clause story_summary exists and if story summary exists then it's a sure thing that story also exists. ( I'm saying this because when  I write `if story is None: raise` block then that block is never executed ).

I hope I answered your question.

Reply:
I'm not sure you have, actually, but I think my original question was wrong because I was confusing story_summaries with stories. Sorry about that.

I think we will need to handle the None case here. This occurs when a story summary is passed in that does not have a corresponding story datastore model. Presumably this is not possible so we should raise an exception if we encounter it.

Reply:
As per the discussion in meeting comment is changed.